plenum/test/batching_3pc/test_batch_rejection.py

```diff
@@ -0,0 +1,97 @@
+import types
+
+import pytest
+
+from stp_core.loop.eventually import eventually
+from plenum.common.constants import DOMAIN_LEDGER_ID
+from plenum.common.util import updateNamedTuple
+from plenum.test.helper import sendRandomRequests, \
+    waitForSufficientRepliesForRequests
+from plenum.test.test_node import getNonPrimaryReplicas, getPrimaryReplica
+
+
+@pytest.fixture(scope="module")
+def setup(tconf, looper, txnPoolNodeSet, client, wallet1):
+    # Patch the 3phase request sending method to send incorrect digest and
+    pr, otherR = getPrimaryReplica(txnPoolNodeSet, instId=0), \
+                 getNonPrimaryReplicas(txnPoolNodeSet, instId=0)
+
+    reqs = sendRandomRequests(wallet1, client, tconf.Max3PCBatchSize)
+    waitForSufficientRepliesForRequests(looper, client, requests=reqs,
+                                        customTimeoutPerReq=tconf.Max3PCBatchWait)
+    stateRoot = pr.stateRootHash(DOMAIN_LEDGER_ID, toHex=False)
+
+    origMethod = pr.create3PCBatch
+    malignedOnce = None
+
+    def badMethod(self, ledgerId):
+        nonlocal malignedOnce
+        pp = origMethod(ledgerId)
+        if not malignedOnce:
+            pp = updateNamedTuple(pp, digest=pp.digest + '123')
+            malignedOnce = True
+        return pp
+
+    pr.create3PCBatch = types.MethodType(badMethod, pr)
+    sendRandomRequests(wallet1, client, tconf.Max3PCBatchSize)
+    return pr, otherR, stateRoot
+
+
+# def testAllDependentBatchesDiscardedAfterBatchReject(setup):
+#     """
+#     Check if a batch if rejected, each batch that was created based on the
+#     rejected batch is discarded.
+#     :return:
+#     """
+
+
+@pytest.fixture(scope="module")
+def reverted(setup, looper):
+    pr, otherR, oldStateRoot = setup
+
+    def chkPps(n):
+        assert len(pr.batches) == n
+
+    def chkStateRoot(root):
+        for r in [pr]+otherR:
+            r.stateRootHash(DOMAIN_LEDGER_ID, toHex=False) == root
+
+    looper.run(eventually(chkPps, 1, retryWait=1, timeout=5))
+    looper.run(eventually(chkStateRoot, oldStateRoot))
+
+
+@pytest.fixture(scope="module")
+def viewChanged(reverted, looper, txnPoolNodeSet):
+    def chk():
+        for n in txnPoolNodeSet:
+            assert n.viewNo == 1
+            assert all([r.primaryName for r in n.replicas])
+
+    looper.run(eventually(chk, retryWait=1, timeout=15))
+
+
+def testTreeStateRevertedAfterBatchRejection(reverted):
+    """"
+    After a batch is rejected, all nodes revert their trees to last known
+    correct state
+    """
+    pass
```

Comment:
Why not, the fixture is a pre-condition, that pre-condition requires certain things to be True for it to be satisfied